ipam/ipam.go

```diff
@@ -0,0 +1,134 @@
+package ipam
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/e2e-harness/pkg/framework"
+	"github.com/giantswarm/e2etests/ipam/provider"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+)
+
+type Config struct {
+	Logger   micrologger.Logger
+	Provider provider.Interface
+}
+
+type IPAM struct {
+	logger   micrologger.Logger
+	provider provider.Interface
+}
+
+func New(config Config) (*IPAM, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.Provider == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Provider must not be empty", config)
+	}
+
+	s := &IPAM{
+		logger:   config.Logger,
+		provider: config.Provider,
+	}
+
+	return s, nil
+}
+
+func (c *IPAM) Test(ctx context.Context) error {
+	var err error
+
+	const (
+		clusterOne   = "cluster0"
+		clusterTwo   = "cluster1"
+		clusterThree = "cluster2"
+		clusterFour  = "cluster3"
+	)
+
+	clusters := []string{clusterOne, clusterTwo, clusterThree}
+
+	c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating three guest clusters: %#v", clusters))
+
+	for _, cn := range clusters {
+		err = c.provider.CreateCluster(cn)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("create guest cluster %s", cn))
+	}
+
+	guestFrameworks := make(map[string]*framework.Guest)
+
+	c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("waiting for three guest clusters to be ready: %#v", clusters))
+
+	for _, cn := range clusters {
+		cfg := framework.GuestConfig{
+			ClusterName: cn,
+			Logger:      c.logger,
+		}
+		guestFramework, err := framework.NewGuest(cfg)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		guestFrameworks[cn] = guestFramework
+		err = guestFramework.Setup()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("guest cluster %s ready", cn))
+	}
+
+	// TODO: Verify subnet properties for three created clusters.
+
+	c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("terminating guest cluster %s and immediately creating new guest cluster %s", clusterTwo, clusterFour))
+
+	c.provider.DeleteCluster(clusterTwo)
+	err = c.provider.CreateCluster(clusterFour)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	{
+		guest := guestFrameworks[clusterTwo]
+		err = guest.WaitForAPIDown()
```

Comment:
Is it important if API is up or down? We should only care about allocated networks.

Reply:
No, not specifically. This was more of a shortcut for waiting VPC to come up / start going down. I might reconsider to change these to actual Subnet/VPC polling, but the problem is that working with E2E tests take already too much time. Your question is reasonable though when considering test execution time.

Reply:
To return here: During development I wished I could do this, but guest VPC info is not exposed and therefore I had to settle with API state.